### PR TITLE
Test feature reader and tica

### PR DESCRIPTION
**Pull request by fabian-paul**
Hi team,
I wrote a kind of "integration test" where I test the feature reader together with Tica. This is a little bit computationally heavy because we need to generate a lot of data to have reliable statistics.
I added the test as a unit test but maybe we should move it somewhere else. Could you help me here?
Fabian

### CODE REVIEW

**Comment by franknoe** [comment]
Thanks a lot! I agree we should separate unit test and computationally 
heavy integration tests, but I am not sure how to do this, because 
integration tests might generally include a lot of data (so git is not 
so good). Perhaps we can still put the tests on github but the data 
elsewhere?
Martin: any ideas?

Am 25/03/15 um 19:00 schrieb fabian-paul:
>
> Hi team,
> I wrote a kind of "integration test" where I test the feature reader 
> together with Tica. This is a little bit computationally heavy because 
> we need to generate a lot of data to have reliable statistics.
> I added the test as a unit test but maybe we should move it somewhere 
> else. Could you help me here?
> Fabian
>
> ------------------------------------------------------------------------
>
>
>         You can view, comment on, or merge this pull request online at:
>
> https://github.com/markovmodel/PyEMMA/pull/146
>
>
>         Commit Summary
>
>   * [tests] joint test of feature reader and tica
>   * Merge branch 'devel' into test_feature_reader_and_tica
>   * [tests] fixed bug in test_featurereader
>
>
>         File Changes
>
>   * *M* pyemma/coordinates/io/feature_reader.py
>     <https://github.com/markovmodel/PyEMMA/pull/146/files#diff-0> (21)
>   * *M* pyemma/coordinates/tests/test_featurereader.py
>     <https://github.com/markovmodel/PyEMMA/pull/146/files#diff-1> (6)
>   * *A* pyemma/coordinates/tests/test_featurereader_and_tica.py
>     <https://github.com/markovmodel/PyEMMA/pull/146/files#diff-2> (84)
>
>
>         Patch Links:
>
>   * https://github.com/markovmodel/PyEMMA/pull/146.patch
>   * https://github.com/markovmodel/PyEMMA/pull/146.diff
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/markovmodel/PyEMMA/pull/146>.
>


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

**Comment by fabian-paul** [comment]
For this particular test, no data is stored in the git folders. All data is generated and stored under /tmp.

**Comment by franknoe** [comment]
Yes, but for other integration tests we will likely be using data and 
therefore we need a general strategy.

There is no clear-cut definition what's a unit or an integration test. 
My feeling is that if more than one API function is tested it is an 
integration test.

Am 25/03/15 um 19:10 schrieb fabian-paul:
>
> For this particular test, no data is stored in the git folders. All 
> data is generated and stored under /tmp.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/markovmodel/PyEMMA/pull/146#issuecomment-86154311>.
>


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

**Comment by marscher** [comment]
Am 25.03.2015 um 19:07 schrieb Frank Noe:
> Thanks a lot! I agree we should separate unit test and computationally
> heavy integration tests, but I am not sure how to do this, because
> integration tests might generally include a lot of data (so git is not
> so good). Perhaps we can still put the tests on github but the data
> elsewhere?
> Martin: any ideas?
I would like to setup a Jenkins job for that, which uses test cases 
stored in git, but local data (group share etc). We decorate those 
tests, so they will not be executed during normal testing.

I nice to have feature would also be a running codespeed (performance 
regression testing) instance, which makes Jenkins obsolete, since its a 
Django application and does not run under Jenkins.

With this architecture we still can maintain our tests in the repo, but 
do not have the mess of storing huge data in it.

**Comment by franknoe** [comment]
I like that.



-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

**Comment by marscher** [comment]
This nose plugin does exactly what we want

http://nose.readthedocs.org/en/latest/plugins/attrib.html

So we only need to setup codespeed and write our heavy (integration) tests.

**Comment by fabian-paul** [comment]
Can I just use this for my test? That is just add
```python
from nose.plugins.attrib import attr
@attr(speed='slow')
```
I guess we are using unittest which is not the same as nosetest. So must we add nodetest to the list of dependencies first?

**Comment by marscher** [comment]
nose is already a testing dependency.

File: pyemma/coordinates/io/feature_reader.py [hunk 1]
```diff
@@ -157,7 +157,13 @@ def dimension(self):
 
         :return:
         """
-        return self.featurizer.dimension()
+        if len(self.featurizer.active_features)==0:
+            # special case: cartesion coordinates
+            top = mdtraj.load(self.topfile)
```
Comment (marscher) [inline pyemma/coordinates/io/feature_reader.py]: @fabian-paul you can reuse the topology which have already been loaded in the featurizer. After this I'm happy to merge your great work!

Comment (marscher) [inline pyemma/coordinates/io/feature_reader.py]: if len(self.featurizer.active_features) == 0:
      return self.featurizer.topology.n_atoms * 3

**Comment by fabian-paul** [comment]
Ok, I have written a test that goes in that direction. First I generate random data and run TICA on it. Then I write out the IC trajectories as ascii files and run TICA on them. From theory we know what the entries of the instantaneous and the time-lagged correlation matrix of the ICs should be. I test this.
The test uses the feature reader, the DataInMemory reader, TICA estimation and TICA projection.

**Comment by marscher** [comment]
very good!

**Comment by franknoe** [comment]
Thanks.

Check out:

https://github.com/franknoe/msmdocs/tree/master/docs/coordinates/lecture_tica/ 
example.ipynb 
<https://github.com/franknoe/msmdocs/blob/master/docs/coordinates/lecture_tica/example.ipynb>

This will make a nice test.

Am 27/03/15 um 16:00 schrieb fabian-paul:
>
> Ok, I have written a test that goes in that direction. First I 
> generate random data and run TICA on it. Then I write out the IC 
> trajectories as ascii files and run TICA on them. From theory we know 
> what the entries of the instantaneous and the time-lagged correlation 
> matrix of the ICs should be. I test this.
> The test uses the feature reader, the DataInMemory reader, TICA 
> estimation and TICA projection.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/markovmodel/PyEMMA/pull/146#issuecomment-86965110>.
>


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

**Pull request merged**